### PR TITLE
Feature/jwt

### PR DESCRIPTION
**Pull request by david-buderus**
### Reason for this PR
- To have a better authentication system that does not require to send every time the password of the usre

### Changes in this PR
- Switch to JWT
- Added login, refresh and logout endpoint
- Every service tests now the authentication by itself
- Clean up dependencies

### CODE REVIEW

File: product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala [hunk 1]
```diff
@@ -19,30 +18,53 @@ trait AuthenticationService extends UC4Service {
...
-  /** Checks if the username and password pair exists */
-  def check(user: String, pw: String): ServiceCall[NotUsed, (String, AuthenticationRole)]
-
   /** Sets the authentication data of a user */
   def setAuthentication(): ServiceCall[AuthenticationUser, Done]
 
   /** Changes the password of the given user */
   def changePassword(username: String): ServiceCall[AuthenticationUser, Done]
 
+  /** Logins a user and return a refresh and a login token in the header */
```
Comment (lrschaef) [inline product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala:27]:
```suggestion
  /** Logs a user in and returns a refresh and a login token in the response header */
```

File: product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala [hunk 2]
```diff
@@ -81,8 +70,136 @@ class AuthenticationServiceImpl(readSide: ReadSide, processor: AuthenticationEve
                 throw new CustomException(code, errorResponse)
             }
         }
-    }(this, ec)
+    }(config, ec)
+
+  /** Logins a user and return a refresh and a login token in the header */
```
Comment (lrschaef) [inline product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala:75]:
```suggestion
  /** Logs a user in and returns a refresh and a login token in the header */
```

File: product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala [hunk 3]
```diff
@@ -19,30 +18,53 @@ trait AuthenticationService extends UC4Service {
...
-        restCall(Method.GET, pathPrefix + "/users?user&pw", check _),
         restCall(Method.POST, pathPrefix + "/users", setAuthentication _),
+
         restCall(Method.PUT, pathPrefix + "/users/:username", changePassword _)(CustomMessageSerializer.jsValueFormatMessageSerializer, MessageSerializer.DoneMessageSerializer),
-        restCall(Method.OPTIONS, pathPrefix + "/users", allowedPut _)
+        restCall(Method.GET, pathPrefix + "/login", login _),
+        restCall(Method.GET, pathPrefix + "/refresh", refresh _),
+        restCall(Method.GET, pathPrefix + "/logout", logout _),
+
+        restCall(Method.OPTIONS, pathPrefix + "/users", allowedPut _),
```
Comment (dhelmert) [inline product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala:53]: The parameter must also be specified in the corresponding OPTIONS call.

```suggestion
        restCall(Method.OPTIONS, pathPrefix + "/users/:username", allowedPut _),
```

File: product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala [hunk 4]
```diff
@@ -81,8 +70,136 @@ class AuthenticationServiceImpl(readSide: ReadSide, processor: AuthenticationEve
...
+    )
+  }
 
   /** Allows PUT */
   override def allowedPut: ServiceCall[NotUsed, Done] = allowedMethodsCustom("PUT")
+
+  /** Allows GET */
+  override def allowedGet: ServiceCall[NotUsed, Done] = allowedMethodsCustom("GET")
+
+  /** Reads username and password out of the header
```
Comment (lrschaef) [inline product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala:183]:
```suggestion
  /** Reads username and password from the header
```

File: product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala [hunk 5]
```diff
@@ -19,30 +18,53 @@ trait AuthenticationService extends UC4Service {
...
   /** Sets the authentication data of a user */
   def setAuthentication(): ServiceCall[AuthenticationUser, Done]
 
   /** Changes the password of the given user */
   def changePassword(username: String): ServiceCall[AuthenticationUser, Done]
 
+  /** Logins a user and return a refresh and a login token in the header */
+  def login: ServiceCall[NotUsed, Done]
+
+  /** Generates a new login token out of a refresh token */
```
Comment (lrschaef) [inline product_code/authentication_service/api/src/main/scala/de/upb/cs/uc4/authentication/api/AuthenticationService.scala:30]:
```suggestion
  /** Generates a new login token from a refresh token */
```

File: product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala [hunk 6]
```diff
@@ -81,8 +70,136 @@ class AuthenticationServiceImpl(readSide: ReadSide, processor: AuthenticationEve
...
+            ("Set-Cookie", s"refresh=$refreshToken; SameSite=Strict; Secure; HttpOnly; Expires=${dateFormat.format(nowRefresh.getTime)} ;;" +
+              s"login=$loginToken; SameSite=Strict; Secure; HttpOnly; Max-Age=${logoutTimer * 60}")
+          )), Done)
+        }
+      case None =>
+        throw CustomException.BasicAuthorizationError
+    }
+  }
+
+  /** Generates a new login token out of a refresh token */
```
Comment (lrschaef) [inline product_code/authentication_service/impl/src/main/scala/de/upb/cs/uc4/authentication/impl/AuthenticationServiceImpl.scala:123]:
```suggestion
  /** Generates a new login token from a refresh token */
```

File: product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationServiceStub.scala [hunk 7]
```diff
@@ -37,4 +20,16 @@ class AuthenticationServiceStub extends AuthenticationService {
 
   /** This Methods needs to allow a GET-Method */
   override def allowedPut: ServiceCall[NotUsed, Done] = ServiceCall { _ => Future.successful(Done) }
+
+  /** Logins a user and return a refresh and a login token in the header */
```
Comment (lrschaef) [inline product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationServiceStub.scala:24]:
```suggestion
  /** Logs a user in and return a refresh and a login token in the header */
```

File: product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationServiceStub.scala [hunk 8]
```diff
@@ -37,4 +20,16 @@ class AuthenticationServiceStub extends AuthenticationService {
...
   /** This Methods needs to allow a GET-Method */
   override def allowedPut: ServiceCall[NotUsed, Done] = ServiceCall { _ => Future.successful(Done) }
+
+  /** Logins a user and return a refresh and a login token in the header */
+  override def login: ServiceCall[NotUsed, Done] = ServiceCall { _ => Future.successful(Done) }
+
+  /** Allows GET */
+  override def allowedGet: ServiceCall[NotUsed, Done] = ServiceCall { _ => Future.successful(Done) }
+
+  /** Generates a new login token out of a refresh token */
```
Comment (lrschaef) [inline product_code/authentication_service/api/src/test/scala/de/upb/cs/uc4/authentication/AuthenticationServiceStub.scala:30]:
```suggestion
  /** Generates a new login token out from a refresh token */
```

**Review by lrschaef — CHANGES_REQUESTED** [review]

**Review by dhelmert — CHANGES_REQUESTED** [review]

**Review by lrschaef — APPROVED** [review]

**Review by dhelmert — APPROVED** [review]

**Pull request merged**